File: day24/day24.cpp
#include <climits>
#include <fstream>
#include <deque>
#include <vector>
#include <unordered_map>

#include <fmt/format.h>

using namespace std;

struct Pos
{
	int x;
	int y;

	bool operator==(const Pos& p) const noexcept
	{
		return x == p.x && y == p.y;
	}
};

struct PosHash
{
	size_t operator()(const Pos& p) const noexcept
	{
		return (hash<int>{}(p.x)<<5)^(hash<int>{}(p.y));
	}
};

struct Location
{
	int value;
	Pos pos;
};

static int distance(const vector<string>& map, Pos start, Pos end)
{
	static const int dx[] = {0, 1, 0, -1};
	static const int dy[] = {-1, 0, 1, 0};
	unordered_map<Pos, int, PosHash> cost;
	deque<Pos> queue;

	queue.push_back(start);
	cost[start] = 0;
	while (!queue.empty())
	{
		Pos pos = queue.front();
		if (pos == end)
		{
			break;
		}
		queue.pop_front();
		for (int i = 0; i < 4; i++)
		{
			Pos np{pos.x+dx[i], pos.y+dy[i]};
			if (map[np.y][np.x] == '#')
			{
				continue;
			}
			if (cost.find(np) != cost.end())
			{
				continue;
			}
			cost[np] = cost[pos] + 1;
			queue.push_back(np);
		}
	}
	return cost[end];
}

static pair<int, int> solve(const vector<string>& map)
{
	// find the locs in the map
	vector<Location> locs;
	for (size_t y = 0; y < map.size(); y++)
	{
		for (size_t x = 0; x < map[y].size(); x++)
		{
			switch (map[y][x])
			{
			case '#':
			case '.':
				break;
			default:
				locs.emplace_back(
					Location{map[y][x], Pos{(int)x, (int)y}}
					);
				break;
			}
		}
	}
	sort(locs.begin(), locs.end(), [](const auto& a, const auto& b) {
		return a.value < b. value;
	});

	// for each location couple find the distance
	vector<vector<int>> dist(locs.size());
	for (auto& d: dist)
	{
		d.resize(locs.size());
	}
	for (size_t i = 0; i < locs.size(); i++)
	{
		for (size_t j = 0; j < i; j++)
		{
			dist[j][i] = dist[i][j] = distance(
				map, locs[i].pos, locs[j].pos);
		}
	}

	// find the length
	vector<int> indices;
	for (size_t i = 1; i < locs.size(); i++)
	{
		indices.push_back(i);
	}

	int part1 = INT_MAX;
	int part2 = INT_MAX;
	do
	{
		int len = dist[0][indices[0]];
		for (size_t i = 1; i < indices.size(); i++)
		{
			len += dist[indices[i-1]][indices[i]];
		}
		if (part1 > len)
		{
			part1 = len;
		}
		len += dist[0][indices.back()];
		if (part2 > len)
		{
			part2 = len;
		}
	} while (next_permutation(indices.begin(), indices.end()));

	return make_pair(part1, part2);
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> map;
	string line;
	while (getline(input, line))
	{
		map.emplace_back(move(line));
	}
	input.close();

	auto [part1, part2] = solve(map);
	fmt::print("Part1: {}\nPart2: {}\n", part1, part2);
	return 0;
}

File: day24/day24.py
#!/usr/bin/env python3

from itertools import permutations
import sys

def distance(mymap, start, end):
    distance = {}
    queue = []
    distance[start] = 0
    queue.append(start)
    while queue:
        pos = queue.pop(0)
        if pos == end:
            break

        for dx, dy in ((0,-1),(1,0),(0,1),(-1,0)):
            np = (pos[0]+dx, pos[1]+dy)
            if mymap[np[1]][np[0]] == "#":
                continue

            if np in distance:
                continue

            distance[np] = distance[pos] + 1
            queue.append(np)

    return distance[end]

def solve(mymap):
    locations = []
    for y, row in enumerate(mymap):
        for x, value in enumerate(row):
            if value != "#" and value != ".":
                locations.append((value, (x, y)))

    locations.sort()

    dist = {}
    for i, (loc1, pos1) in enumerate(locations):
        for j in range(i):
            loc2, pos2 = locations[j]
            dist[i, j] = dist[j, i] = distance(mymap, pos1, pos2)

    part1 = 1e12
    part2 = 1e12
    for path in permutations(range(1, len(locations))):
        length = dist[0, path[0]]
        for x in range(len(path)-1):
            length += dist[path[x], path[x+1]]
        if part1 > length:
            part1 = length
        length += dist[0, path[-1]]
        if part2 > length:
            part2 = length

    return part1, part2

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    mymap = [list(x) for x in txt.split("\n")]
    part1, part2 = solve(mymap)
    print("Part1:", part1)
    print("Part2:", part2)

